### PR TITLE
fix merge with timestamp (#571)

### PR DESCRIPTION
**Pull request by seyguai**
## Summary
The merge algorithm isn't valid when timestamp is set: it only removes from the existing dataset, not the incoming one.

**Related Issue: #571**

**PR Type: Bug Fix**

**Status: Ready to review**

**Breaking change? No**

### Solution
New algorithm :
- identifies the keys that have duplicated rows from both the incoming & the existing dataset
- extract the rows that match those keys
- sort those rows using the timestamp column, keep all but the latest entry for each windows (the unwanted duplicates)
- remove those rows from the union of the two datasets

### How has this been tested?
Updating an existing unit test.

## Contributor checklist:
Go over all the following points, and put an `x` in all the boxes that apply.
If you're unsure about any of these, don't hesitate to ask. We're here to help!
- [x] My code follows the code style of this project.
- [x] I have updated the Release notes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala [hunk 1]
```diff
@@ -809,40 +808,54 @@ trait IngestionJob extends SparkJob {
...
-
     // We remove from the incoming data the records to delete
     val updatesDF = merge.delete
       .map(condition => in.filter(s"not ($condition)"))
       .getOrElse(in)
+
+    val (toDeleteDF, mergedDF) = merge.timestamp
+      .map { timestamp =>
+        // We only keep the first occurrence of each record, from both datasets
+        val orderingWindow =
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala:819]: We order the records by having the last record first

**Review by hayssams — APPROVED** [review]
LGTM

File: src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala [hunk 2]
```diff
@@ -809,40 +808,54 @@ trait IngestionJob extends SparkJob {
...
+          .join(keysWithDuplicatesDF, merge.key)
+          .select(allRowsDF.columns.map(col): _*)
+
+        val duplicatedRowsDF = rowsWithDuplicatesDF
+          .withColumn("rownum", row_number.over(orderingWindow))
+          .where(col("rownum") =!= 1)
+          .drop("rownum")
+
+        (duplicatedRowsDF, allRowsDF.except(duplicatedRowsDF))
+      }
```
Comment (abdelhakimbendjabeur) [inline src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala:839]: Why not just do 
```scala
val (toDeleteDF, mergedDF) = merge.timestamp
  .map { timestamp =>
    // Put everything in one DF
    val allRowsDF = existing.union(updatesDF)
  
    // Define window
    val orderingWindow =
          Window.partitionBy(merge.key.head, merge.key.tail: _*).orderBy(col(timestamp).desc)

    // Deduplicate
    val mergedDF= allRowsDF
      .withColumn("rownum", row_number.over(orderingWindow ))
      .where(col("rownum") === 1)
      .drop("rownum")

    // Compute rows that will be deleted for logging purposes apparently
    val toDeleteDF = rowsWithDuplicatesDF
      .withColumn("rownum", row_number.over(orderingWindow))
      .where(col("rownum") =!= 1) 
      .drop("rownum")

     (toDeleteDF, mergedDF)
  }
```

Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala:839]: LGTM

**Review by hayssams — APPROVED** [review]

File: src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala (hunk 2 above)
Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/ingest/IngestionJob.scala:839]: @seyguai Feel free to merge anytime

**Pull request merged**